### DISCUSSION TITLE
What does this mod do?

### DISCUSSION THREAD

**Discussion by k0nstruct** [Steveplays28/noisium#2] [General]
I'm pretty curious about this mod, and its performance gains, what does it do (in layman's terms)?

I've seen your explanation of replacing functions to optimize blockstate stuff, so I guess it's about filling in the chunk info faster, and not computing the structures themselves?

It's a new mod, so it's unpopular, and I love testing out new work! Especially ones uncalled for, so your project has me really curious :D. Do you have future plans for other optimizations?

**Comment by wicivo** [comment]
> I'm pretty curious about this mod, and its performance gains, what does it do (in layman's terms)?
> 
> I've seen your explanation of replacing functions to optimize blockstate stuff, so I guess it's about filling in the chunk info faster, and not computing the structures themselves?
> 
> It's a new mod, so it's unpopular, and I love testing out new work! Especially ones uncalled for, so your project has me really curious :D. Do you have future plans for other optimizations?

I agree with you, and also is there any sense to use this mod together with c2me

**Reply by Steveplays28** [reply]
Hi wicivo,

Noisium is fully compatible with C2ME, and I can recommend these mods together!
In my testing they've worked together really well, so far.

**Comment by Steveplays28** [comment]
Hey owosoup,

Noisium changes some world generation functions that other mods don't touch, to fill in the gaps left by other performance optimisation mods.
Most notably, `NoiseChunkGenerator#populateNoise` is optimised to speed up blockstate placement when generating new chunks. There are also 3 other smaller optimisations, that increase biome population speed, chunk unlocking speed, and the speed of sampling blockstates for generation.
In `NoiseChunkGenerator#populateNoise`, setting the blockstate via abstractions/built-in functions is bypassed. Instead, the blocks are set directly in the palette storage, thus bypassing a lot of calculations and things Minecraft does that are normally useful when blocks are set, but when generating the world only slow it down.

**Reply by k0nstruct** [reply]
Oh wow, that's a good call. I hope this mod doesn't get incompatible when other optimization mods (like c2me) start tampering with it as well. I'll benchmark it when I get home!

**Reply by Steveplays28** [reply]
Awesome, thanks. I made sure not to step in C2ME's way, it should be fully compatible with Noisium.

**Reply by Steveplays28** [reply]
Also, to answer your other questions, structure generation hasn't been touched yet. Might try to do that in the future!
I want to do more profiling and find out what else needs work when it comes to worldgen performance.

I've only really talked about this mod in my Discord server and in [Distant Horizons' Discord server](https://modrinth.com/mod/distanthorizons). I'm going to announce the release today, since it got approved on Modrinth today.

Looking forward to seeing your benchmarks.

**Reply by k0nstruct** [reply]
This is very interesting; without this mod, I get bad fps using sodium loading at 12 threads, even with deffered chunk updates. The loading is fast, yes, but I get spikes and overall lower fps.

WITH the mod however, I don't need to use less threads to get my fps back, I can use 12 threads just fine and it rolls out chunks with way less fps tanking! it's like it allows sodium to do more with deferred chunk updates. You were on the money for the ~30% uplift of performance, it's what I'm getting with generation speeds. It works great!

Thank you, really, this adresses the biggest issue of my modpack yet, I was using bobby to fake chunks but now I can use 32 render distance just fine, albeit with some remaining fps spikes that are irrelevant since I don't notice them. 

_I'm getting amazing performance while flying around in creative! Thank you for your work, kudos!_
![2023-11-01_20 28 39](https://github.com/Steveplays28/noisium/assets/95828007/208d031b-3aff-4ed9-82b7-8dd81f2e8298)

**Reply by Steveplays28** [reply]
That's amazing! Had no idea the impact could be that massive.
Just tested with Sodium and some other performance mods, I also notice a huge speedup and less lag with high chunk build thread counts.

Thanks for the kind words!

**Reply by ghost** [reply]
Why exactly does Noisium improve the performance of chunk updates as a worldgen optimization mod?

**Reply by Steveplays28** [reply]
I think it might be that worldgen blocks the main thread less since it runs faster, causing a decrease in stuttering when loading new chunks. Still need to do more investigating though.

**Reply by AnOpenSauceDev** [reply]
I decided to try benchmarking the `sample()` method, and got around 11% faster noise sampling compared to vanilla's logic. I'm actually quite surprised that taking apart the loop and directly sampling seems to be so much faster.

**Reply by UltraToon** [reply]
when you claimed that this mod was compatible with datapack generation mods, is it also compatible with terralith and tectonic? also with yungs suite?

**Reply by Steveplays28** [reply]
Hi UltraToon, Noisium is fully compatible with Yung's, Terralith, and Tectonic.

**Comment by Phoenix-Starlight** [comment]
Just a question, does replacing the for loop syntax actually bring any benefits?

**Reply by Steveplays28** [reply]
Hey Phoenix-Starlight, changing the loop syntax does bring some benefits due to the way Java/the JVM works. Enhanced `for` loops are a lot worse, performance wise, compared to `fori` loops.

Thanks for checking out Noisium!

**Comment by Phoenix-Starlight** [comment]
I should be more specific, does benchmarking (with it being JIT'd) it
compared to enhanced for show an improvement? (I'm looking more at
concrete results)

**Comment by ajaxnotmeta** [comment]
just a quick question, does this thing compatibles to world gen datapacks like terralith or even william wythers biomes mod/datapack?
especially Wythers Mod version as it was somewhat weirder because you need terrablender to use it alongside terralith mod variant while the datapacks somewhat works just fine using both

**Reply by AnOpenSauceDev** [reply]
Based off of what I've seen in the code, there should be no issues with generation, alongside 1:1 worldgen, based off of how the optimizations work.

**Reply by Steveplays28** [reply]
Hi ajaxnotmeta, as anOpenSauceDev said, this mod is fully compatible with both worldgen datapacks and mods.
There's 1:1 parity in the worldgen.

**Comment by liangmoe** [comment]
How do i know noisium is necessary for optimization or not ingame? I see no differences with/without it by viewing mspt and tps, see no difference while chunk generating too, ive read the recent discussion but that is not enough for me as a casual player, my os is win10, tested on forge 47.1.3 with modernfix and radium

**Reply by Steveplays28** [reply]
Hi liangmoe, the best way to check if it makes a difference is to do a few benchmarks using VisualVM/Spark.
The speedup is dependent on how fast your hardware is. This mod makes the most difference on low-end/mid-end hardware.

**Reply by liangmoe** [reply]
This is my spark, what ive done was kill spawn chunk with modernfix, and use freecam to see the whole test was done or not and run spark when i immediately landed and ofc use the same seed, maybe im not good at reading spark

[without noisium](https://spark.lucko.me/I5sXyBYIdF), [with noisium](https://spark.lucko.me/nurtcugSKX)

**Reply by Steveplays28** [reply]
It seems like the server is 2% more idle in these tests, see `MinecraftServer.waitUntilNextTick()`.
Not sure if that's random variation or actual improvements, you'll have to run more identical tests to verify that.

**Reply by liangmoe** [reply]
Yes, but the problem is your mod page description describe that noisium speed up 20%-30% when generating new chunks in vanilla minecraft, and since we play modded minecraft, it is impossible to throw away lithium (or radium), in that case it is not actually a improvemend that you can barely see differences and you still mention it, minecraft commuinty is not smart, they can only read descrpiton and add the mod, usually they dont have a proper way to "benchmark" minecraft, if pepo add mod that is not causing crash they just thought they added another mod with optimization mod without any problem, and they just spread suggestion that noisium is improving world gen speed with their modlist, people just belive in you from your description withou knowing or benchmarking anything, this is not good, and most likely it just baiting and just spread infinity

**Reply by Steveplays28** [reply]
I've got an open issue for improving the benchmarking results in the readme,
What do you mean by throwing away Lithium? I added Lithium compatibility to Noisium recently.

**Reply by liangmoe** [reply]
Im not meaning of noisium cannot play with lithium, im mean play lithium+noisium cant see the differences that uve described in your mod page description - speed up 20%-30% when generating new chunks in vanilla minecraft

**Reply by Steveplays28** [reply]
As said before, it's quite hardware-dependent at the moment, maybe your system is fast enough that it doesn't make a big difference.

**Comment by GGCristo** [comment]
By "performance" do you mean it is faster? or it is better optimized (a.k.a consumes less resources)? I think C2ME fits in the first category consuming more resources, in order to give you more "performance", but it can be very harsh in old hardware, I want to run a minecraft server in an old laptop, so I am more interested in less resources usage
